presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java

```diff
@@ -1463,6 +1469,74 @@ protected Node aggregateResult(Node aggregate, Node nextResult)
         throw new UnsupportedOperationException("not yet implemented");
     }
 
+    private static String decodeString(SqlBaseParser.UnicodeStringLiteralContext context)
+    {
+        char escape;
+        if (context.UESCAPE() != null) {
+            String escapeString = unquote(context.STRING().getText());
+            check(!escapeString.isEmpty(), "Empty Unicode escape character", context);
+            check(escapeString.length() == 1, "Invalid Unicode escape character: " + escapeString, context);
+            escape = escapeString.charAt(0);
+            check(isValidUnicodeEscape(escape), "Invalid Unicode escape character: " + escapeString, context);
+        }
+        else {
+            escape = '\\';
+        }
+
+        String rawContent = unquote(context.UNICODE_STRING().getText().substring(2));
+        StringBuilder builder = new StringBuilder();
+        int beginCodeIndex = 0;
+        int charactersNeeded = 0;
+        for (int i = 0; i < rawContent.length(); i++) {
```

Comment:
Honestly, this loop is structured in a quite messy way.

You kind of wanted to write it in a streaming fashion, but you didn't do that. The code looks ahead and looks back. Ideally, it should do neither.

Alternatively, you could choose to not do streaming at all, which would be pretty clean and easy to read for something like this.

If you are willing to restructure it, I'm totally for it. Otherwise, given the localized nature of this function, I'm ok with the current shape.